File: tests/js/cohere/test_reference.ts
/**
 * Conformance test: Reference instrumentation for Cohere.
 */
import { trace } from "@opentelemetry/api";
import { flushAndShutdownOtel, setupOtel } from "../otel";

const nodeProcess = globalThis.process as NodeJS.Process & {
  env: Record<string, string | undefined>;
  exit(code?: number): never;
};

const MOCK_BASE_URL = nodeProcess.env.MOCK_LLM_URL!;
const tracer = trace.getTracer("gen_ai.reference");

async function main() {
  console.log("=== Reference: Cohere JS Conformance Test ===");
  const otel = setupOtel();

  const { CohereClient } = await import("cohere-ai");
  const client = new CohereClient({ token: "mock-key", baseUrl: MOCK_BASE_URL });

  // Scenario: chat
  console.log("  [chat] basic chat completion (reference)");
  await tracer.startActiveSpan("chat command-r-plus", async (span) => {
    const requestModel = "command-r-plus";
    span.setAttribute("gen_ai.operation.name", "chat");
    span.setAttribute("gen_ai.provider.name", "cohere");
    span.setAttribute("gen_ai.request.model", requestModel);
    const resp = await client.chat({
      model: requestModel,
      message: "Say hello.",
    });
    if ((resp as any).generationId) {
      span.setAttribute("gen_ai.response.id", (resp as any).generationId);
    }
    if ((resp as any).finishReason) {
      span.setAttribute("gen_ai.response.finish_reasons", [(resp as any).finishReason]);
    }
    if ((resp as any).meta?.billedUnits?.inputTokens) {
      span.setAttribute("gen_ai.usage.input_tokens", (resp as any).meta.billedUnits.inputTokens);
    }
    if ((resp as any).meta?.billedUnits?.outputTokens) {
      span.setAttribute("gen_ai.usage.output_tokens", (resp as any).meta.billedUnits.outputTokens);
    }
    console.log(`    -> ${resp.text.slice(0, 60)}`);
    span.end();
  });

  // Scenario: embeddings
  console.log("  [embeddings] embedding generation (reference)");
  await tracer.startActiveSpan("embeddings embed-english-v3.0", async (span) => {
    const requestModel = "embed-english-v3.0";
    span.setAttribute("gen_ai.operation.name", "embeddings");
    span.setAttribute("gen_ai.provider.name", "cohere");
    span.setAttribute("gen_ai.request.model", requestModel);
    const resp = await client.embed({
      model: requestModel,
      texts: ["Hello, world!"],
      inputType: "search_document",
    });
    if ((resp as any).meta?.billedUnits?.inputTokens) {
      span.setAttribute("gen_ai.usage.input_tokens", (resp as any).meta.billedUnits.inputTokens);
    }
    const embeddings = resp.embeddings as number[][];
    console.log(`    -> embedding dim: ${embeddings[0].length}`);
    span.end();
  });

  await flushAndShutdownOtel(otel);
}

main().catch((e) => { console.error(e); nodeProcess.exit(1); });

File: tests/python/mistralai/test_reference.py
"""Conformance test: reference instrumentation for Mistral AI.

Exercises: chat, chat_streaming, embeddings
against a mock server, with manual OTel spans.
"""

import os

from otel_setup import setup_otel, flush_and_shutdown
from opentelemetry import trace

MOCK_BASE_URL = os.environ["MOCK_LLM_URL"]

_reference_tracer = trace.get_tracer("gen_ai.reference")


def run_chat(client):
    """Scenario: basic chat completion with reference instrumentation."""
    print("  [chat] basic chat completion (reference)")
    request_model = "mistral-large-latest"
    with _reference_tracer.start_as_current_span("chat mistral-large-latest") as span:
        span.set_attribute("gen_ai.operation.name", "chat")
        span.set_attribute("gen_ai.provider.name", "mistral")
        span.set_attribute("gen_ai.request.model", request_model)
        resp = client.chat.complete(
            model=request_model,
            messages=[{"role": "user", "content": "Say hello."}],
        )
        if resp.model:
            span.set_attribute("gen_ai.response.model", resp.model)
        if resp.id:
            span.set_attribute("gen_ai.response.id", resp.id)
        if resp.choices:
            span.set_attribute("gen_ai.response.finish_reasons",
                               [c.finish_reason for c in resp.choices if c.finish_reason])
        if resp.usage:
            span.set_attribute("gen_ai.usage.input_tokens", resp.usage.prompt_tokens)
            span.set_attribute("gen_ai.usage.output_tokens", resp.usage.completion_tokens)
        print(f"    -> {resp.choices[0].message.content[:60]}")


def run_chat_streaming(client):
    """Scenario: streaming chat completion with reference instrumentation."""
    print("  [chat_streaming] streaming chat completion (reference)")
    request_model = "mistral-large-latest"
    with _reference_tracer.start_as_current_span("chat mistral-large-latest") as span:
        span.set_attribute("gen_ai.operation.name", "chat")
        span.set_attribute("gen_ai.provider.name", "mistral")
        span.set_attribute("gen_ai.request.model", request_model)
        text = ""
        stream = client.chat.stream(
            model=request_model,
            messages=[{"role": "user", "content": "Tell me a joke."}],
        )
        for event in stream:
            if event.data.choices and event.data.choices[0].delta.content:
                text += event.data.choices[0].delta.content
        print(f"    -> {text[:60]}")


def run_embeddings(client):
    """Scenario: embedding generation with reference instrumentation."""
    print("  [embeddings] embedding generation (reference)")
    request_model = "mistral-embed"
    with _reference_tracer.start_as_current_span("embeddings mistral-embed") as span:
        span.set_attribute("gen_ai.operation.name", "embeddings")
        span.set_attribute("gen_ai.provider.name", "mistral")
        span.set_attribute("gen_ai.request.model", request_model)
        resp = client.embeddings.create(
            model=request_model,
            inputs=["Hello, world!"],
        )
        if resp.usage:
            span.set_attribute("gen_ai.usage.input_tokens", resp.usage.prompt_tokens)
        print(f"    -> embedding dim: {len(resp.data[0].embedding)}")


def main():
    print("=== Reference: Mistral AI Conformance Test ===")

    tp, lp, mp = setup_otel()
    # NO instrument() call – reference instrumentation only

    from mistralai.client import Mistral
    client = Mistral(
        api_key="mock-key",
        server_url=MOCK_BASE_URL,
    )

    run_chat(client)
    try:
        run_chat_streaming(client)
    except Exception as e:
        print(f"    WARNING: streaming failed: {e}")

    run_embeddings(client)

    flush_and_shutdown(tp, lp, mp)


if __name__ == "__main__":
    main()
